### PR TITLE
cellSaveData: funcFile improvements, some error checks added

### LINKED ISSUE #7738: Haze [BLUS30094] regression as of 0.0.9-9791
## Quick summary
An error message prevents you from continuing to the main menu now.

## Details
Haze was in a playable state until 0.0.9-9791, since then if you try to start the game you'll be met by this error message: 

image

It'll prompt you to delete the save data it just created, once you're done you back out and are met with the same error message because it created a save once again that it thinks is corrupt. No way around this error message.

I went through builds and found that the game was working fine all the way up to 0.0.9-9790, after that every version of RPCS3 going forward has had this issue. I've attached logs from both builds.

[RPCS3 0.0.9-9790 (good).log](https://github.com/RPCS3/rpcs3/files/4306741/RPCS3.0.0.9-9790.good.log)
[RPCS3 0.0.9-9791 (bad).log](https://github.com/RPCS3/rpcs3/files/4306742/RPCS3.0.0.9-9791.bad.log)

### PR DESCRIPTION
**Pull request by elad335**
* Add more logging, simplify code to use std::string subfunctions and strcpy_trunc and add one more error condition.
* Allow '_' at filenames start and extension.
* Check if reading offset is valid, fix error code to CELL_SAVEDATA_ERROR_FAILURE.
* Don't create empty file on error of write ops.
* Don't allow "." and ".." filenames on funcFile, return CELL_SAVEDATA_ERROR_BROKEN.
* Check listSet->fixedListNum.
* Check listSet->fixedList, listSet->focusDirName,  listSet->newData->dirName for nullptr and its directory items names.
* Check listSet->newData->iconPosition.
* Check statSet->setParam->parental_level for old sdk.
* Return an error if listSet->focusPosition is NEWDATA and listSet->newData is nullptr.
* Simplify savedata directory list selection.
* Fixed savedata lock after fmt::throw_exception.
* Filter directory lists to allowed savedata directories.

Fixes #7738.

### CODE REVIEW

**Comment by ZachFett** [comment]
Logs from Haze thread, as requested:
[RPCS3.0.0.9-9790.good.log](https://github.com/RPCS3/rpcs3/files/4310205/RPCS3.0.0.9-9790.good.log)
[RPCS3.0.0.9-9791.bad.log](https://github.com/RPCS3/rpcs3/files/4310206/RPCS3.0.0.9-9791.bad.log)

**Comment by elad335** [comment]
Added some changes based on provate hw tests.

**Comment by elad335** [comment]
Added some error checks for fun.

File: rpcs3/Emu/Cell/Modules/cellSaveData.cpp [hunk 1]
```diff
@@ -1400,64 +1523,87 @@ static NEVER_INLINE error_code savedata_op(ppu_thread& ppu, u32 operation, u32 v
 				break;
 			}
 
-			const auto fileName = fileSet->fileName.get_ptr();
+			const char* fileName = fileSet->fileName.get_ptr();
 
-			const auto termpos = std::find(fileName, fileName + CELL_SAVEDATA_FILENAME_SIZE, '\0');
-
-			if (termpos == fileName + CELL_SAVEDATA_FILENAME_SIZE)
+			if (const auto termpos = std::memchr(fileName, '\0', CELL_SAVEDATA_FILENAME_SIZE); termpos)
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/Modules/cellSaveData.cpp]: Redundant "termpos" after declaration

**Comment by elad335** [comment]
* Don't allow "." and ".." filenames on funcFile, return CELL_SAVEDATA_ERROR_FAILURE

File: rpcs3/Emu/Cell/Modules/cellSaveData.cpp [hunk 2]
```diff
@@ -89,7 +90,10 @@ namespace
 
 vm::gvar<savedata_context> g_savedata_context;
 
-std::mutex g_savedata_mutex;
+struct savedata_mutex
+{
+	std::mutex mutex;
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/Modules/cellSaveData.cpp]: std::mutex may not support it... It needs a semaphore<> because it can be left in any state.

**Comment by isshininu** [comment]
This PR also fixed same problem with saving in The Evil Within.

**Pull request merged**